src/tls/certificate_manager.rs

```diff
@@ -78,8 +91,72 @@ impl CertificateManager {
         }
     }
 
-    pub fn reload_from_directory(&mut self, directory: PathBuf) -> Result<(), Error> {
-        let certificates =  try!(create_records_from_directory(&directory));
+    pub fn get_or_generate_self_signed_certificate(&self, hostname: String) -> io::Result<CertificateRecord> {
+
+        // This is blocking - because we generate the result and write off thread - and at the end
+        // of that operation we know whether it succeeded, we can't borrow self to add the
+        // CertificateRecord for it in the other thread easily.  This kind of needs to be blocking
+        // anyway, we can't proceed without the resulting value anyway in most uses.
+
+        if let Some(certificate_record) = self.get_certificate(hostname.clone()) {
+            info!("Using existing self-signed cert for {}", hostname);
+            return Ok(certificate_record);
+        }
+
+        let mut directory = self.directory.clone();
+        directory.push(hostname.clone());
+
+        info!("Generating new self-signed cert for {}", hostname);
+        let generator = X509Generator::new()
+            .set_bitlength(2048)
+            .set_valid_period(365 * 2)
+            .add_name("CN".to_owned(), hostname.clone())
+            .set_sign_hash(Type::SHA256);
+
+        let gen_result = generator.generate();
+        if let Ok((cert, pkey)) = gen_result {
+
+            // Write the cert and pkey PEM files
+
+            // Ensure the directory exists
+            let _ = fs::create_dir_all(directory.clone());
```

Comment:
Or you can do something similar to https://github.com/fxbox/foxbox/blob/master/src/profile_service.rs#L61